### PR TITLE
hidpp20: add new profile layout (0x04)

### LINKED ISSUE #774: Logitech G903 Hero Profile layout not supported
Hi using version 0.9.906 on arch so i have a device file for the g903 hero, output from ratbagd:

```
ratbag error: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO: error opening hidraw node (Operation not supported)
ratbag debug: New device: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: device assigned driver hidpp20
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc091
ratbag debug: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO is device '/dev/hidraw7'.
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc091
ratbag debug: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO is device '/dev/hidraw8'.
ratbag debug: HID report ID 01
ratbag debug: HID report ID 03
ratbag debug: HID report ID 04
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc091
ratbag debug: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO is device '/dev/hidraw9'.
ratbag debug: HID report ID 10
ratbag debug: HID report ID 11
ratbag raw: hidpp write: 11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 18 04 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 00 08 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 01 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0001 is at 0x01
ratbag raw: hidpp write: 11 ff 01 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 08 1d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 01 00 01 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 02 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 03 00 02 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 1d 4b 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 06 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 10 01 00 02 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 07 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 80 71 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 81 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 09 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 81 10 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0a 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 80 60 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0b 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 22 01 00 01 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0c 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 00 c2 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 02 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0e 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 03 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 0f 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 05 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 10 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 06 60 04 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 11 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 11 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 12 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 30 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 13 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 90 60 04 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 14 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 91 60 04 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 15 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 a1 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 16 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 01 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 17 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 b1 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 1d f3 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 19 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 1e 00 40 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 1a 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 1e b0 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 1b 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 18 63 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 1c 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 1e 22 60 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp write: 11 ff 01 18 1d 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 01 18 21 21 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: 'Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO' is using protocol v4.2
ratbag raw: 'Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO' has 30 features
ratbag raw: Init feature HIDPP_PAGE_ROOT (0x0000) 
ratbag raw: hidpp write: 11 ff 00 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 00 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0000 is at 0x00
ratbag raw: Init feature HIDPP_PAGE_FEATURE_SET (0x0001) 
ratbag raw: hidpp write: 11 ff 00 08 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 01 00 01 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0001 is at 0x01
ratbag raw: Init feature HIDPP_PAGE_DEVICE_INFO (0x0003) 
ratbag raw: hidpp write: 11 ff 00 08 00 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 02 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0003 is at 0x02
ratbag raw: unknown feature 0x0003
ratbag raw: Init feature HIDPP_PAGE_DEVICE_NAME (0x0005) 
ratbag raw: hidpp write: 11 ff 00 08 00 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 03 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0005 is at 0x03
ratbag raw: unknown feature 0x0005
ratbag raw: Init feature HIDPP_PAGE_WIRELESS_DEVICE_STATUS (0x1d4b) 
ratbag raw: hidpp write: 11 ff 00 08 1d 4b 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 04 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x1d4b is at 0x04
ratbag raw: unknown feature 0x1d4b
ratbag raw: Init feature HIDPP_PAGE_RESET (0x0020) 
ratbag raw: hidpp write: 11 ff 00 08 00 20 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 05 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x0020 is at 0x05
ratbag raw: unknown feature 0x0020
ratbag raw: Init feature HIDPP_PAGE_BATTERY_VOLTAGE (0x1001) 
ratbag raw: hidpp write: 11 ff 00 08 10 01 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 06 00 02 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x1001 is at 0x06
ratbag raw: hidpp write: 10 ff 06 08 00 00 00
ratbag raw: hidpp read:  11 ff 06 08 10 63 80 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag debug: device battery voltage is 4195mv, status 80 
ratbag raw: Init feature 0x8071 (0x8071) 
ratbag raw: hidpp write: 11 ff 00 08 80 71 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 07 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x8071 is at 0x07
ratbag raw: unknown feature 0x8071
ratbag raw: Init feature HIDPP_PAGE_ONBOARD_PROFILES (0x8100) 
ratbag raw: hidpp write: 11 ff 00 08 81 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  11 ff 00 08 08 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: feature 0x8100 is at 0x08
ratbag debug: device has onboard profiles
ratbag raw: hidpp write: 10 ff 08 08 00 00 00
ratbag raw: hidpp read:  11 ff 08 08 01 04 01 05 01 0c 10 00 ff 0a 04 00 00 00 00 00
ratbag error: Profile layout not supported: 0x04.
ratbag error: Logitech G903 LIGHTSPEED Wireless Gaming Mouse w/ HERO: error opening hidraw node (Operation not supported)
```

And Also 

```
ratbag debug: New device: Logitech G903 LS
ratbag debug: Using data directory '/usr/share/libratbag'
ratbag debug: device assigned driver hidpp20
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc539
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0xffffc539
ratbag debug: hidraw info: bus 0x03 vendor 0x46d product 0x4087
ratbag debug: Logitech G903 LS is device '/dev/hidraw15'.
ratbag debug: HID report ID 01
ratbag debug: HID report ID 0e
ratbag debug: HID report ID 02
ratbag debug: HID report ID 10
ratbag debug: HID report ID 11
ratbag debug: HID report ID 20
ratbag debug: HID report ID 21
ratbag raw: hidpp write: 11 ff 00 18 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
ratbag raw: hidpp read:  10 01 8f 00 18 09 00
ratbag debug:     HID++ error from the device (1): ERR_UNSUPPORTED (09)
```

### PR DESCRIPTION
**Pull request by FFY00**
Fixes #774

### CODE REVIEW

File: src/hidpp20.h [hunk 1]
```diff
@@ -699,7 +699,8 @@ struct hidpp20_profile {
 	uint16_t dpi[HIDPP20_DPI_COUNT];
 	union hidpp20_button_binding buttons[32];
 	union hidpp20_macro_data *macros[32];
-	struct hidpp20_led leds[HIDPP20_LED_COUNT];
+	struct hidpp20_led leds_8070[HIDPP20_LED_COUNT];
+	struct hidpp20_led leds_8071[HIDPP20_LED_COUNT];
```
Comment (whot) [inline src/hidpp20.h]: i don't quite understand why we need to duplicate this here - are there mice that have both features? if not, then we don't need the same structs duplicated, we can switch at runtime where necessary.

Comment (FFY00) [reply src/hidpp20.h]: This struct maps directly to the memory we read from the mouse. We could instead have `struct hidpp20_led leds[HIDPP20_LED_COUNT * 2];` and then have a define for the index, which would be `0` for 0x8070 and `2` for 0x8071, but I think this way is cleaner.

Comment (FFY00) [reply src/hidpp20.h]: I misread the documentation. It seems this 2 new leds describe the effects triggered when the device goes to power save mode.

**Comment by FFY00** [comment]
@whot Ready for review.

**Comment by FFY00** [comment]
0x8070 has been deprecated in favor of 0x8071. This new feature introduces the concept of low power effects. For each led we have an alternative effect that is triggered in low power mode (device goes to sleep). This patch only declares those leds, it does not allow us to configure them. @whot do we want to support this?

**Comment by whot** [comment]
I'd say no, we don't. at least not at this point, we have enough tricky other features that we don't properly support and exposing this is going to be flimsy at best. If anything, we could hack this in the driver that we automatically pick a low-power equivalent mode where appropriate. But I don't think we need to expose this in the UI

**Pull request merged**